### PR TITLE
Bt google acl write

### PR DESCRIPTION
**Pull request by vivster7**
ACL write permissions should work would google. Lots of changes made here, so lets do a group code review tomorrow morning before pulling it in.

### CODE REVIEW

**Comment by bradtaylor** [comment]
Could we squash these commits, after addressing any additional questions from code review?  I see some WIP commits I made, etc.

File: src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/AuthorizationProvider.scala [hunk 1]
```diff
@@ -2,22 +2,53 @@ package org.broadinstitute.dsde.agora.server.dataaccess.acls
 
 import org.broadinstitute.dsde.agora.server.model.AgoraEntity
 
-trait AuthorizationProvider {
-  private def filterAuthorizedEntityByReadPermissions(authEntity: AuthorizedAgoraEntity): Option[AgoraEntity] = {
-    if(authEntity.authorization.canRead) authEntity.entity else None
-  }
-
-  def authorizationsForEntity(agoraEntity: Option[AgoraEntity], username: String): AuthorizedAgoraEntity
+import scala.concurrent.Future
```
Comment (bradtaylor) [inline src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/AuthorizationProvider.scala]: Is Future used in this class?

File: src/main/scala/org/broadinstitute/dsde/agora/server/model/GoogleAccessControls.scala [hunk 2]
```diff
@@ -1,20 +1,53 @@
...
+import org.joda.time.DateTime
+
 case class ProjectTeam(projectNumber: Option[String], team: Option[String])
 
+object GoogleAccessControls {
+  def bucketKind = "storage#bucketAccessControl"
+  def objectKind = "storage#objectAccessControl"
+}
+
+//case class BucketAccessControls(kind:        Option[String] = Option("storage#bucketAccessControl"),
```
Comment (bradtaylor) [inline src/main/scala/org/broadinstitute/dsde/agora/server/model/GoogleAccessControls.scala]: Will these be removed before merge?

File: src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/AuthorizationProvider.scala [hunk 3]
```diff
@@ -2,22 +2,53 @@ package org.broadinstitute.dsde.agora.server.dataaccess.acls
...
+
+  /**
+   * Given a sequence of AgoraEntities, return a new sequence containing only those a given user is authorized to read
+   * @param entities The sequence of entities to filter
+   * @param username The user whose permissions you wish to filter by
+   * @return Seq[AgoraEntity], filtered by user permissions
+   */
+  def filterByReadPermissions(entities: Seq[AgoraEntity], username: String): Seq[AgoraEntity] =
+    entities
+      .filter(entity => isAuthorizedForRead(entity, username))
```
Comment (bradtaylor) [inline src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/AuthorizationProvider.scala]: We are still calling Await in each call to entityAuthorization, so I am still worried about parallelism. Looking forward to learning in group code review :)

File: src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala [hunk 4]
```diff
@@ -16,77 +16,155 @@ import spray.http.StatusCodes._
...
+  override def namespaceAuthorization(entity: AgoraEntity, username: String): AgoraPermissions = {
+    val bucketNotFound = checkBucketExistence(entity, username)
 
-  val timeout = 20 seconds
+    if (bucketNotFound) return new AgoraPermissions(Create)
+    else getBucketPermissions(entity, username)
+  }
 
-  val gcsBaseUrl = "https://www.googleapis.com/storage/v1/b/"
+      private def checkBucketExistence(entity: AgoraEntity, username: String): Boolean = {
```
Comment (bradtaylor) [inline src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala]: These 2 functions seem to have an extra indent

File: src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala [hunk 5]
```diff
@@ -16,77 +16,155 @@ import spray.http.StatusCodes._
...
+      if (r.status != OK) throw new Exception("Could not add user ACL to object.")
     }
   }
 
-  def getGcsBucketRoleRequestUrl(namespace: String, userEmail: String): String = gcsBaseUrl + namespaceToBucketName(namespace) + "/acl/user-" + userEmail
+      private def createBucket(entity: AgoraEntity, username: String, bucket: Promise[HttpResponse]): Unit = {
+        val url = createBucketUrl
+        val content = BucketResource(name = Option(bucketName(entity)))
+        val response = gcsPOST(url, content)
+        response onSuccess { case r => bucket.success(r) }
```
Comment (bradtaylor) [inline src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala]: what happens on Failure?

File: src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala [hunk 6]
```diff
@@ -16,77 +16,155 @@ import spray.http.StatusCodes._
...
+    val bucket = Promise[HttpResponse]()
+    val bucketAcl = Promise[HttpResponse]()
+    val _object = Promise[HttpResponse]()
+    val objectAcl = Promise[HttpResponse]()
 
-  def entityToObjectName(entity: AgoraEntity): String = "agoraEntity_" + entity.entityType.get + "_" + entity.name.get + "_" + entity.snapshotId.get
+    createBucket(entity, username, bucket)
+
+    bucket.future onSuccess { case r =>
+      if (r.status != OK && r.status != Conflict) throw new Exception("Could not create bucket.")
```
Comment (bradtaylor) [inline src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala]: AddHandler does not seem to be looking for these exceptions, in order to complete request. Do they crash the system?

File: src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala [hunk 7]
```diff
@@ -16,77 +16,155 @@ import spray.http.StatusCodes._
...
 
-object GcsAuthorizationProvider extends AuthorizationProvider with LazyLogging {
-  implicit val actorSystem = ActorSystem("agora")
-  import actorSystem.dispatcher
+  override def namespaceAuthorization(entity: AgoraEntity, username: String): AgoraPermissions = {
+    val bucketNotFound = checkBucketExistence(entity, username)
 
-  val timeout = 20 seconds
+    if (bucketNotFound) return new AgoraPermissions(Create)
+    else getBucketPermissions(entity, username)
```
Comment (dshiga) [inline src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala]: Are we cool with this format for if statements? Do we prefer using match?

File: src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala [hunk 8]
```diff
@@ -16,77 +16,155 @@ import spray.http.StatusCodes._
...
+
+  override def createEntityAuthorizations(entity: AgoraEntity, username: String): Unit = {
+    val bucket = Promise[HttpResponse]()
+    val bucketAcl = Promise[HttpResponse]()
+    val _object = Promise[HttpResponse]()
+    val objectAcl = Promise[HttpResponse]()
 
-  def entityToObjectName(entity: AgoraEntity): String = "agoraEntity_" + entity.entityType.get + "_" + entity.name.get + "_" + entity.snapshotId.get
+    createBucket(entity, username, bucket)
+
```
Comment (bradtaylor) [inline src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala]: I think future users might appreciate a helpful comment here about the flow of calls

File: src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala [hunk 9]
```diff
@@ -2,91 +2,180 @@
...
-        throw new Exception("Trying to get Google ACLs for AgoraEntity with missing namespace, name, or snapshotId")
+  def doesBucketExist(entity: AgoraEntity, username: String): Boolean = {
+    val url = bucketUrl(entity)
+    val bucket = gcsGET(url)
+
+    val bucketNotFound = bucket map { r: HttpResponse => r.status == NotFound }
+
+    bucket.onFailure { case _ => throw GoogleConnectionException() }
+
+    val isbucketFound = !Await.result(bucketNotFound, timeout)
```
Comment (jacarey) [inline src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala]: isBucketFound

File: src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala [hunk 10]
```diff
@@ -2,91 +2,180 @@
...
+  def addUserAclToObject(entity: AgoraEntity, username: String): Unit = {
+    val url = createObjectAclUrl(entity)
+    val content = GoogleAccessControls(kind   = Option(GoogleAccessControls.objectKind),
+                                       entity = Option(prefixUser(username)),
+                                       role   = Option(GcsRole.Owner))
+    val response = gcsPOST(url, content)
+
+    response.onSuccess { case r: HttpResponse => if (r.status != OK) throw GoogleConnectionException() }
+    response.onFailure { case t: Throwable => throw GoogleConnectionException() }
+  }
```
Comment (jacarey) [inline src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala]: We are calling post on url and content and handling success/failure in three separate places. Seems like we could consolidate this logic.

**Comment by jacarey** [comment]
Broken tests ?

**Pull request merged**